enhancements/cvo/object-removal-manifest-annotation.md

```diff
@@ -0,0 +1,144 @@
+
+### Non-Goals
+
+* This enhancement does not add any new objects for managment by CVO, therefore the new delete annotation only applies to the currently supported manifests and any that may be added in the future.
+
+## Proposal
+
+When the following annotation appears in a CVO supported manifest and is set to "true" the associated object will be removed from the cluster by the CVO.
+Values other than `true` will result in a CVO error and no further processing of the manifest.
+
+```
+apiVersion: apps/v1
+...
+metadata:
+...
+  annotations:
+    release.openshift.io/delete: "true"
+```
+The existing CVO ordering scheme defined [here](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#what-is-the-order-that-resources-get-createdupdated-in) will also be used for object removal.
+This both minimizes CVO code changes and provides a simple method of deleting multiple objects by reversing the order in which they were created.
+
+### User Stories
+
+#### Story 1
+
+Remove the cluster-autoscaler-operator deployment.
+The existing cluster-autoscaler-operator deployment manifest 0000_50_cluster-autoscaler-operator_07_deployment.yaml is modified to contain the delete annotation:
+```
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: cluster-autoscaler-operator
+  namespace: openshift-machine-api
+  annotations:
+    release.openshift.io/delete: "true"
+...
+```
+Additional manifest properties such as `spec` may be set if convenient (e.g. because you are looking to make a minimal change vs. a previous version of the manifest), but those properties have no affect on manifests with the delete annotation.
+
+#### Story 2
+
+In release 4.5 two jobs, openshift-service-catalog-controller-manager-remover and openshift-service-catalog-apiserver-remover, were created to remove the Service Catalog.
+For release 4.6 these jobs and all their supporting cluster objects should be removed.
+This enhancement allows a user to accomplish this in a very simple manner by modifying all the manifests associated with these two jobs to contain the delete annotation and changing the `run level` portion and the number portion, if one exists, of each manifest filename to reverse their create order.
+In this case the original manifests used for creation are:
+* 0000_50_cluster-svcat-apiserver-operator_04_roles.yaml
+* 0000_50_cluster-svcat-apiserver-operator_05_serviceaccount.yaml
+* 0000_50_cluster-svcat-controller-manager-operator_06_roles.yaml
+* 0000_50_cluster-svcat-controller-manager-operator_07_serviceaccount.yaml
+* 0000_90_cluster-svcat-apiserver-operator_01_remover_job.yaml
+* 0000_90_cluster-svcat-controller-manager-operator_01_remover_job.yaml
+
+To achieve correct delete ordering the manifests are renamed to:
+* 0000_50_cluster-svcat-controller-manager-operator_01_remover_job.yaml
+* 0000_50_cluster-svcat-apiserver-operator_01_remover_job.yaml
```

Comment:
Can we remove the namespace they run in, and implicitly remove the jobs that way?  Or do they live in a namespace which needs to survive?

Reply:
My point was to show that one could just reverse the order of the create if they wanted to remove everything. However I did miss the manifest that originally created the namespace so I'll add it. And given that the namespace is `openshift-service-catalog-removed` it seems to have only been created for these remover jobs.